### PR TITLE
feat: add no-std support

### PR DESCRIPTION
**Pull request by anonrig**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -12,7 +12,7 @@ description = "Fast WHATWG Compliant URL parser"
 documentation = "https://docs.rs/ada-url"
 readme = "README.md"
 keywords = ["url", "parser", "whatwg", "performance"]
-categories = ["parser-implementations", "web-programming", "encoding"]
+categories = ["parser-implementations", "web-programming", "encoding", "no-std::no-alloc", "no-std"]
```
Comment (KSXGitHub) [inline Cargo.toml:15]: Categories are not keywords.

Valid categories here: https://crates.io/categories

Can you use `cargo publish --dry-run --allow-dirty` to check if your Cargo.toml is valid.

Comment (anonrig) [reply Cargo.toml]: Yes, these are all valid. The correct list is available at https://crates.io/category_slugs

File: src/lib.rs [hunk 2]
```diff
@@ -30,17 +30,27 @@
...
 
 #[cfg(feature = "serde")]
 extern crate serde;
 
+#[cfg(feature = "alloc")]
+extern crate alloc;
+
+#[cfg(feature = "std")]
+extern crate std;
+
```
Comment (KSXGitHub) [inline src/lib.rs:53]: This shouldn't be necessary.

File: src/lib.rs [hunk 3]
```diff
@@ -30,17 +30,27 @@
 //! ```toml
 //! ada-url = { version = "1", features = ["serde"] }
 //! ```
+#![no_std]
```
Comment (KSXGitHub) [inline src/lib.rs:33]: This should be a `cfg(feature)`

Comment (KSXGitHub) [inline src/lib.rs:33]: Should've been this

```suggestion
#![cfg_attr(not(feature = "alloc"), no_std)]
```

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
There are many more needed changes, but I'm too sleepy for now. Wait for my PR tomorrow.

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
Merge https://github.com/ada-url/rust/pull/42 instead.

**Comment by anonrig** [comment]
Closing in favor of @KSXGitHub's pull request.

**Pull request closed without merging**